File: cwms-data-api/src/main/java/cwms/cda/data/dto/locationlevel/LocationLevel.java
/*
 *
 * MIT License
 *
 * Copyright (c) 2025 Hydrologic Engineering Center
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE
 * SOFTWARE.
 */

package cwms.cda.data.dto.locationlevel;

import java.math.BigDecimal;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonRootName;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;
import cwms.cda.data.dto.CwmsDTO;
import cwms.cda.data.dto.CwmsDTOValidator;
import cwms.cda.formatters.Formats;
import cwms.cda.formatters.UnsupportedFormatException;
import cwms.cda.formatters.annotations.FormattableWith;
import cwms.cda.formatters.json.JsonV1;
import cwms.cda.formatters.json.JsonV2;
import cwms.cda.formatters.xml.XMLv2;
import io.swagger.v3.oas.annotations.media.Schema;

import hec.data.level.ILocationLevelRef;
import hec.data.level.JDomLocationLevelImpl;

@JsonRootName("LocationLevel")
@JsonDeserialize(builder = LocationLevel.Builder.class)
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.KebabCaseStrategy.class)
@FormattableWith(contentType = Formats.JSONV2, formatter = JsonV2.class, aliases = {Formats.DEFAULT, Formats.JSON})
@FormattableWith(contentType = Formats.JSONV1, formatter = JsonV1.class)
@FormattableWith(contentType = Formats.XMLV2, formatter = XMLv2.class, aliases = {Formats.XML})
@JsonIgnoreProperties(ignoreUnknown = true)
public abstract class LocationLevel extends CwmsDTO {
    @JsonProperty(required = true)
    @Schema(description = "Name of the location level")

    private final String locationLevelId;
    @Schema(description = "Timeseries ID (e.g. from the times series catalog) to use as the "
            + "location level. Mutually exclusive with seasonalValues and "
            + "siParameterUnitsConstantValue")

    private final String specifiedLevelId;

    private final ZonedDateTime expirationDate;

    @Schema(description = "Data Type such as Stage, Elevation, or others.")

    private final String parameterId;
    @Schema(description = "To indicate if single or aggregate value",
            allowableValues = {"Inst", "Ave", "Min", "Max", "Total"})

    private final String parameterTypeId;
    @Schema(description = "Indicating whether or not to interpolate between seasonal values.",
            allowableValues = {"T", "F"})

    private final String interpolateString;

    @Schema(description = "Units the provided levels are in")

    private final String levelUnitsId;
    @Schema(description = "The date/time at which this location level configuration takes effect.")
    @JsonFormat(shape = JsonFormat.Shape.STRING)

    private final ZonedDateTime levelDate;

    private final String levelComment;
    @Schema(description = "0 if parameterTypeId is Inst. Otherwise duration indicating the time "
            + "window of the aggregate value.")

    private final String durationId;

    private final BigDecimal attributeValue;

    private final String attributeUnitsId;

    private final String attributeParameterTypeId;

    private final String attributeParameterId;

    private final String attributeDurationId;

    private final String attributeComment;

    LocationLevel(LocationLevel.Builder builder) {
        super(builder.officeId);
        specifiedLevelId = builder.specifiedLevelId;
        parameterTypeId = builder.parameterTypeId;
        levelUnitsId = builder.levelUnitsId;
        levelDate = builder.levelDate;
        levelComment = builder.levelComment;
        durationId = builder.durationId;
        attributeValue = builder.attributeValue;
        attributeUnitsId = builder.attributeUnitsId;
        attributeParameterTypeId = builder.attributeParameterTypeId;
        attributeParameterId = builder.attributeParameterId;
        attributeDurationId = builder.attributeDurationId;
        attributeComment = builder.attributeComment;
        locationLevelId = builder.locationId;
        parameterId = builder.parameterId;
        interpolateString = builder.interpolateString;
        expirationDate = builder.expirationDate;
    }

    public String getSpecifiedLevelId() {
        return specifiedLevelId;
    }

    public String getParameterTypeId() {
        return parameterTypeId;
    }

    public String getLevelUnitsId() {
        return levelUnitsId;
    }

    public ZonedDateTime getLevelDate() {
        return levelDate;
    }

    public ZonedDateTime getExpirationDate() {
        return expirationDate;
    }

    public String getLevelComment() {
        return levelComment;
    }

    public String getInterpolateString() {
        return interpolateString;
    }

    public String getDurationId() {
        return durationId;
    }

    public BigDecimal getAttributeValue() {
        return attributeValue;
    }

    public String getAttributeUnitsId() {
        return attributeUnitsId;
    }

    public String getAttributeParameterTypeId() {
        return attributeParameterTypeId;
    }

    public String getAttributeParameterId() {
        return attributeParameterId;
    }

    public String getAttributeDurationId() {
        return attributeDurationId;
    }

    public String getAttributeComment() {
        return attributeComment;
    }

    public String getLocationLevelId() {
        return locationLevelId;
    }

    public String getParameterId() {
        return parameterId;
    }

    @JsonPOJOBuilder
    @JsonNaming(PropertyNamingStrategies.KebabCaseStrategy.class)
    public abstract static class Builder<T extends Builder<T>> {
        String specifiedLevelId;
        String parameterTypeId;
        String parameterId;
        String levelUnitsId;
        ZonedDateTime levelDate;
        String levelComment;
        String durationId;
        BigDecimal attributeValue;
        String attributeUnitsId;
        String attributeParameterTypeId;
        String attributeParameterId;
        String attributeDurationId;
        String attributeComment;
        String locationId;
        String officeId;
        String interpolateString;
        ZonedDateTime expirationDate;
        final Map<String, Consumer<Object>> propertyFunctionMap = new HashMap<>();

        @JsonCreator
        protected Builder(@JsonProperty(value = "location-level-id", required = true) String name,
                @JsonProperty(value = "level-date", required = true) ZonedDateTime lvlDate) {
            locationId = name;
            levelDate = lvlDate;
        }

        Builder(LocationLevel copyFrom) {
            withAttributeComment(copyFrom.getAttributeComment());
            withAttributeDurationId(copyFrom.getAttributeDurationId());
            withAttributeParameterId(copyFrom.getAttributeParameterId());
            withLocationLevelId(copyFrom.getLocationLevelId());
            withAttributeValue(copyFrom.getAttributeValue());
            withAttributeParameterTypeId(copyFrom.getAttributeParameterTypeId());
            withAttributeUnitsId(copyFrom.getAttributeUnitsId());
            withDurationId(copyFrom.getDurationId());
            withLevelComment(copyFrom.getLevelComment());
            withLevelDate(copyFrom.getLevelDate());
            withLevelUnitsId(copyFrom.getLevelUnitsId());
            withOfficeId(copyFrom.getOfficeId());
            withParameterTypeId(copyFrom.getParameterTypeId());
            withSpecifiedLevelId(copyFrom.getSpecifiedLevelId());
        }

        Builder(JDomLocationLevelImpl copyFrom) {
            withAttributeComment(copyFrom.getAttributeComment());
            withAttributeDurationId(copyFrom.getAttributeDurationId());
            withAttributeParameterId(copyFrom.getAttributeParameterId());
            ILocationLevelRef locationLevelRef = copyFrom.getLocationLevelRef();
            if (locationLevelRef != null) {
                withLocationLevelId(locationLevelRef.getLocationLevelId());
            }
            withAttributeValue(copyFrom.getAttributeValue());
            withAttributeParameterTypeId(copyFrom.getAttributeParameterTypeId());
            withAttributeUnitsId(copyFrom.getAttributeUnitsId());
            withDurationId(copyFrom.getDurationId());
            withLevelComment(copyFrom.getLevelComment());
            Date copyLevelDate = copyFrom.getLevelDate();
            if (copyLevelDate != null) {
                withLevelDate(ZonedDateTime.ofInstant(copyLevelDate.toInstant(), ZoneId.of("UTC")));
            }
            withLevelUnitsId(copyFrom.getLevelUnitsId());
            withOfficeId(copyFrom.getOfficeId());
            withParameterId(copyFrom.getParameterId());
            withParameterTypeId(copyFrom.getParameterTypeId());
            withSpecifiedLevelId(copyFrom.getSpecifiedLevelId());
        }

        protected T self() {
            return (T) this;
        }

        @JsonIgnore
        public T withProperty(String propertyName, Object value) {
            Consumer<Object> function = propertyFunctionMap.get(propertyName);
            if (function == null) {
                throw new IllegalArgumentException("Property Name does not exist for Location "
                        + "Level");
            }
            function.accept(value);
            return self();
        }

        public T withSpecifiedLevelId(String specifiedLevelId) {
            this.specifiedLevelId = specifiedLevelId;
            return self();
        }

        public T withParameterTypeId(String parameterTypeId) {
            this.parameterTypeId = parameterTypeId;
            return self();
        }

        public T withParameterId(String parameterId) {
            this.parameterId = parameterId;
            return self();
        }

        public T withExpirationDate(ZonedDateTime expirationDate) {
            this.expirationDate = expirationDate;
            return self();
        }

        public T withLevelUnitsId(String levelUnitsId) {
            this.levelUnitsId = levelUnitsId;
            return self();
        }

        public T withLevelDate(ZonedDateTime levelDate) {
            this.levelDate = levelDate;
            return self();
        }

        public T withLevelComment(String levelComment) {
            this.levelComment = levelComment;
            return self();
        }

        public T withInterpolateString(String interpolateString) {
            this.interpolateString = interpolateString;
            return self();
        }

        public T withDurationId(String durationId) {
            this.durationId = durationId;
            return self();
        }

        public T withAttributeValue(BigDecimal attributeValue) {
            this.attributeValue = attributeValue;
            return self();
        }

        public T withAttributeUnitsId(String attributeUnitsId) {
            this.attributeUnitsId = attributeUnitsId;
            return self();
        }

        public T withAttributeParameterTypeId(String attributeParameterTypeId) {
            this.attributeParameterTypeId = attributeParameterTypeId;
            return self();
        }

        public T withAttributeParameterId(String attributeParameterId) {
            this.attributeParameterId = attributeParameterId;
            return self();
        }

        public T withAttributeDurationId(String attributeDurationId) {
            this.attributeDurationId = attributeDurationId;
            return self();
        }

        public T withAttributeComment(String attributeComment) {
            this.attributeComment = attributeComment;
            return self();
        }

        public T withLocationLevelId(String locationId) {
            this.locationId = locationId;
            return self();
        }

        public T withOfficeId(String officeId) {
            this.officeId = officeId;
            return self();
        }
    }

    public static LocationLevel getUpdatedLocationLevel(LocationLevel existingLevel,
            LocationLevel updatedLevel, ZonedDateTime unmarshalledDate) {

        String specifiedLevelId = (updatedLevel.getSpecifiedLevelId() == null
                ? existingLevel.getSpecifiedLevelId() : updatedLevel.getSpecifiedLevelId());
        String parameterTypeId = (updatedLevel.getParameterTypeId() == null
                ? existingLevel.getParameterTypeId() : updatedLevel.getParameterTypeId());
        String parameterId = (updatedLevel.getParameterId() == null
                ? existingLevel.getParameterId() : updatedLevel.getParameterId());

        String levelUnitsId = (updatedLevel.getLevelUnitsId() == null
                ? existingLevel.getLevelUnitsId() : updatedLevel.getLevelUnitsId());
        String levelComment = (updatedLevel.getLevelComment() == null
                ? existingLevel.getLevelComment() : updatedLevel.getLevelComment());

        String durationId = (updatedLevel.getDurationId() == null
                ? existingLevel.getDurationId() : updatedLevel.getDurationId());
        BigDecimal attributeValue = (updatedLevel.getAttributeValue() == null
                ? existingLevel.getAttributeValue() : updatedLevel.getAttributeValue());
        String attributeUnitsId = (updatedLevel.getAttributeUnitsId() == null
                ? existingLevel.getAttributeUnitsId() : updatedLevel.getAttributeUnitsId());
        String attributeParameterTypeId = (updatedLevel.getAttributeParameterTypeId() == null
                ? existingLevel.getAttributeParameterTypeId() :
                updatedLevel.getAttributeParameterTypeId());
        String attributeParameterId = (updatedLevel.getAttributeParameterId() == null
                ? existingLevel.getAttributeParameterId() : updatedLevel.getAttributeParameterId());
        String attributeDurationId = (updatedLevel.getAttributeDurationId() == null
                ? existingLevel.getAttributeDurationId() : updatedLevel.getAttributeDurationId());
        String attributeComment = (updatedLevel.getAttributeComment() == null
                ? existingLevel.getAttributeComment() : updatedLevel.getAttributeComment());
        String locationId = (updatedLevel.getLocationLevelId() == null
                ? existingLevel.getLocationLevelId() : updatedLevel.getLocationLevelId());
        String officeId = (updatedLevel.getOfficeId() == null
                ? existingLevel.getOfficeId() : updatedLevel.getOfficeId());

        if (existingLevel.getAttributeValue() == null) {
            attributeUnitsId = null;
        }

        LocationLevel.Builder builder = null;

        if (existingLevel instanceof VirtualLocationLevel) {
            VirtualLocationLevel virtualLevel = (VirtualLocationLevel) existingLevel;
            VirtualLocationLevel updatedVirtualLevel = (VirtualLocationLevel) updatedLevel;

            String constituentConnections = (updatedVirtualLevel.getConstituentConnections() == null
                    ? virtualLevel.getConstituentConnections() : updatedVirtualLevel.getConstituentConnections());
            List<VirtualLocationLevel.RatingConstituent> constituents = updatedVirtualLevel.getConstituents() == null
                    ? virtualLevel.getConstituents() : updatedVirtualLevel.getConstituents();

            builder = new VirtualLocationLevel.Builder(locationId, unmarshalledDate)
                    .withExpirationDate(updatedVirtualLevel.getExpirationDate() == null
                        ? virtualLevel.getExpirationDate() : updatedVirtualLevel.getExpirationDate())
                    .withConstituents(constituents)
                    .withConstituentConnections(constituentConnections);
        } else if (existingLevel instanceof ConstantLocationLevel) {
            ConstantLocationLevel constantLevel = (ConstantLocationLevel) existingLevel;
            ConstantLocationLevel updatedConstantLevel = (ConstantLocationLevel) updatedLevel;

            Double siParameterUnitsConstantValue = (updatedConstantLevel.getConstantValue() == null
                    ? constantLevel.getConstantValue() : updatedConstantLevel.getConstantValue());

            builder = new ConstantLocationLevel.Builder(locationId, unmarshalledDate)
                    .withConstantValue(siParameterUnitsConstantValue);
        } else if (existingLevel instanceof TimeSeriesLocationLevel) {
            TimeSeriesLocationLevel timeSeriesLevel = (TimeSeriesLocationLevel) existingLevel;
            TimeSeriesLocationLevel updatedTimeSeriesLevel = (TimeSeriesLocationLevel) updatedLevel;

            String seasonalTimeSeriesId = (updatedTimeSeriesLevel.getSeasonalTimeSeriesId() == null
                    ? timeSeriesLevel.getSeasonalTimeSeriesId() : updatedTimeSeriesLevel.getSeasonalTimeSeriesId());

            builder = new TimeSeriesLocationLevel.Builder(locationId, unmarshalledDate, seasonalTimeSeriesId);
        } else if (existingLevel instanceof SeasonalLocationLevel) {
            SeasonalLocationLevel seasonalLevel = (SeasonalLocationLevel) existingLevel;
            SeasonalLocationLevel updatedSeasonalLevel = (SeasonalLocationLevel) updatedLevel;

            ZonedDateTime intervalOrigin = (updatedSeasonalLevel.getIntervalOrigin() == null
                    ? seasonalLevel.getIntervalOrigin() : updatedSeasonalLevel.getIntervalOrigin());
            Integer intervalMinutes = (updatedSeasonalLevel.getIntervalMinutes() == null
                    ? seasonalLevel.getIntervalMinutes() : updatedSeasonalLevel.getIntervalMinutes());
            Integer intervalMonths = (updatedSeasonalLevel.getIntervalMonths() == null
                    ? seasonalLevel.getIntervalMonths() : updatedSeasonalLevel.getIntervalMonths());
            String interpolateString = (updatedSeasonalLevel.getInterpolateString() == null
                    ? seasonalLevel.getInterpolateString() : updatedSeasonalLevel.getInterpolateString());

            List<SeasonalValueBean> seasonalValues = (updatedSeasonalLevel.getSeasonalValues() == null
                    ? seasonalLevel.getSeasonalValues() : updatedSeasonalLevel.getSeasonalValues());

            if (seasonalLevel.getIntervalMonths() != null && seasonalLevel.getIntervalMonths() > 0) {
                intervalMinutes = null;
            } else if (seasonalLevel.getIntervalMinutes() != null
                    && seasonalLevel.getIntervalMinutes() > 0) {
                intervalMonths = null;
            }

            builder = new SeasonalLocationLevel.Builder(locationId, unmarshalledDate)
                    .withSeasonalValues(seasonalValues)
                    .withIntervalMinutes(intervalMinutes)
                    .withIntervalMonths(intervalMonths)
                    .withIntervalOrigin(intervalOrigin)
                    .withInterpolateString(interpolateString);
        }
        if (builder == null) {
            throw new UnsupportedFormatException("Unsupported Location Level type: "
                    + existingLevel.getClass().getName());
        }

        builder.withParameterTypeId(parameterTypeId)
                .withSpecifiedLevelId(specifiedLevelId)
                .withParameterId(parameterId)
                .withLevelUnitsId(levelUnitsId)
                .withLevelComment(levelComment)
                .withDurationId(durationId)
                .withAttributeValue(attributeValue)
                .withAttributeUnitsId(attributeUnitsId)
                .withAttributeParameterTypeId(attributeParameterTypeId)
                .withAttributeParameterId(attributeParameterId)
                .withAttributeDurationId(attributeDurationId)
                .withAttributeComment(attributeComment)
                .withOfficeId(officeId);

        if (builder instanceof SeasonalLocationLevel.Builder) {
            return ((SeasonalLocationLevel.Builder) builder).build();
        } else if (builder instanceof TimeSeriesLocationLevel.Builder) {
            return ((TimeSeriesLocationLevel.Builder) builder).build();
        } else if (builder instanceof ConstantLocationLevel.Builder) {
            return ((ConstantLocationLevel.Builder) builder).build();
        } else if (builder instanceof VirtualLocationLevel.Builder) {
            return ((VirtualLocationLevel.Builder) builder).build();
        } else {
            throw new UnsupportedFormatException("Unsupported Location Level type: "
                    + existingLevel.getClass().getName());
        }
    }

    @Override
    protected void validateInternal(CwmsDTOValidator validator) {
        super.validateInternal(validator);
        validator.required(getOfficeId(), "office-id");
        validator.required(getLocationLevelId(), "location-level-id");
        validator.required(getLevelDate(), "level-date");
    }
}

File: cda-etl/src/cda_etl/location_level.py
#  MIT License
#  Copyright (c) 2026 Hydrologic Engineering Center
#  Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to deal
#  in the Software without restriction, including without limitation the rights
#  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#  copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#  The above copyright notice and this permission notice shall be included in all
#  copies or substantial portions of the Software.
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
#  SOFTWARE.
import logging
import time
from datetime import datetime
from typing import Iterable

import cwms
import utils.filesystem_store as filesystem_store
import utils.cda_errors as cda_errors
import utils.log_util as log_util
import utils.threading_util as threading_util
from config import LocationLevelConfig

logger = logging.getLogger(__name__)
# Filename-safe, for the staged path. Display comes from log_util.
DATE_TIME_FORMAT = "%Y-%m-%d %H.%M.%S"
LEVELS_FOLDER = "LocationLevels"

_NO_VALUES = "with no values"
_NO_RECORDS = "with no records to publish"

_tally = log_util.Tally()


def _start_batch() -> log_util.Tally:
    global _tally
    _tally = log_util.Tally()

    return _tally


def _label(work_item) -> str:
    if work_item[4]:
        return f"{work_item[1]} [POR]"

    return f"{work_item[1]} [{log_util.window(work_item[2], work_item[3])}]"


def stage_location_levels(
    office_id: str,
    levels: Iterable[LocationLevelConfig],
    default_start: str | None,
    default_end: str | None,
) -> None:
    levels = list(levels)
    work_items = _build_location_level_work_items(
        office_id, levels, default_start, default_end
    )
    if not work_items:
        logger.debug("No location levels configured for office %s; nothing to extract.", office_id)
        return

    tally = _start_batch()
    started = time.monotonic()
    threading_util.execute_tasks(_download_one_location_level, work_items, label=_label, tally=tally)
    log_util.outcome(
        logger,
        action="Staged",
        noun="location level",
        total=len(work_items),
        tally=tally,
        office_id=office_id,
        elapsed=time.monotonic() - started,
    )


def publish_staged_location_levels(
    office_id: str,
    levels: Iterable[LocationLevelConfig],
    default_start: str | None,
    default_end: str | None,
) -> None:
    levels = list(levels)
    work_items = _build_location_level_work_items(
        office_id, levels, default_start, default_end
    )
    if not work_items:
        logger.debug("No location levels configured for office %s; nothing to load.", office_id)
        return

    tally = _start_batch()
    started = time.monotonic()
    threading_util.execute_tasks(_upload_one_location_level, work_items, label=_label, tally=tally)
    log_util.outcome(
        logger,
        action="Published",
        noun="location level",
        total=len(work_items),
        tally=tally,
        office_id=office_id,
        elapsed=time.monotonic() - started,
    )


def _download_one_location_level(work_item: list[object]) -> None:
    office_id = work_item[0]
    level_id = work_item[1]
    begin = work_item[2]
    end = work_item[3]
    por = work_item[4]

    if por:
        logger.info("Extracting location levels (POR) for %s in office %s", level_id, office_id)
        try:
            level_data = cwms.get_location_levels(level_id_mask=level_id, office_id=office_id).json
        except Exception as error:
            if not cda_errors.is_no_data(error):
                raise

            logger.debug(
                "No location level values for %s in office %s; nothing staged.",
                level_id,
                office_id,
            )
            _tally.record(_NO_VALUES, level_id)
            return

        filesystem_store.write_json(level_data, office_id, LEVELS_FOLDER, f"{level_id}.por")
        return

    logger.info(
        "Extracting location level %s for office %s [%s]",
        level_id,
        office_id,
        log_util.window(begin, end),
    )

    try:
        level_data = cwms.get_location_levels(
            level_id_mask=level_id,
            office_id=office_id,
            begin=begin,
            end=end,
        ).json
    except Exception as error:
        if not cda_errors.is_no_data(error):
            raise

        logger.debug(
            "No location level values for %s in office %s between %s and %s; nothing staged.",
            level_id,
            office_id,
            log_util.display(begin),
            log_util.display(end),
        )
        _tally.record(_NO_VALUES, level_id)
        return

    filesystem_store.write_json(level_data, office_id, LEVELS_FOLDER, level_id)


def _upload_one_location_level(work_item: list[object]) -> None:
    office_id = work_item[0]
    level_id = work_item[1]
    por = work_item[4]

    staged_data = filesystem_store.read_json(
        office_id,
        LEVELS_FOLDER,
        f"{level_id}.por" if por else level_id,
    )
    if staged_data is None:
        raise FileNotFoundError(
            "No staged location level data found."
        )

    levels = staged_data.get("levels", []) if isinstance(staged_data, dict) else []
    if not levels:
        logger.debug("No location level records to publish for %s in office %s", level_id, office_id)
        _tally.record(_NO_RECORDS, level_id)
        return

    for level_record in levels:
        cwms.store_location_level(level_record)


def _build_location_level_work_items(
    office_id: str,
    levels: Iterable[LocationLevelConfig],
    default_start: str | None,
    default_end: str | None,
) -> list[list[object]]:
    work_items: list[list[object]] = []

    for level in levels:
        level_id = level.id
        por = level.period_of_record
        begin = None if por else _parse_timestamp(level.start_time or default_start, "start")
        end = None if por else _parse_timestamp(level.end_time or default_end, "end")
        work_items.append([office_id, level_id, begin, end, por])

    return work_items


def _parse_timestamp(value: str | None, label: str) -> datetime:
    if value is None:
        raise ValueError(f"Missing {label} time for location level processing.")

    normalized = value.strip()
    if normalized.lower() == "now":
        return datetime.now()

    try:
        return datetime.fromisoformat(normalized)
    except ValueError as exc:
        try:
            return datetime.strptime(normalized, "%Y-%m-%d")
        except ValueError:
            raise ValueError(f"Invalid {label} time '{value}'. Use ISO-8601 or YYYY-MM-DD.") from exc


__all__ = ["publish_staged_location_levels", "stage_location_levels"]
